### PR TITLE
Don't push a new frame in `eval ()`

### PR DESCRIPTION
**Pull request by PossiblyAShrub**
This test captures the effect of this change:

```
#### eval with vars follows same scoping as without
proc local-scope {
  var myVar = "foo"
  eval (^(echo $myVar), vars={ someOtherVar: "bar" })
  eval (^(echo $myVar))
}

# In global scope
var myVar = "baz"
eval (^(echo $myVar), vars={ someOtherVar: "bar" })
eval (^(echo $myVar))

local-scope
## STDOUT:
baz
baz
foo
foo
## END
```

### CODE REVIEW

File: core/state.py [hunk 1]
```diff
@@ -1161,14 +1162,33 @@ def __enter__(self):
...
 
         if self.pos_args is not None:
             self.mem.argv_stack.pop()
 
         if self.dollar0 is not None:
             self.mem.SetLocalName(self.dollar0_lval, value.Undef)
 
+    # Note: _Push and _Pop are separate methods because the C++ translation
+    # doesn't like when they are inline in __init__ and __exit__.
+    def _Push(self, pairs):
```
Comment (andychu) [inline core/state.py:1175]: How about creating free functions like:

```
state.PushVars(mem, restore)
state.PopVars(mem, restore)
```

and then using it from both `ctx_Shvar` and `ctx_Eval` ?

---

Also, I think where this is going is that we are going to move it up to the YSH level:


```
push-vars ({a: 42, b: 'foo'} {
  echo $a $b
}
```

There is a thread about "unifying stacks with context managers" on Zulip

**Review by andychu — COMMENTED** [review]
Great thanks!  Glad it was easy to find this

And I'm glad it came up at the same time as the Hay bug ...

File: core/state.py (hunk 1 above)
Comment (PossiblyAShrub) [reply core/state.py:1175]: What about using `ctx_Eval` instead of `ctx_Shvar` for `shvar`?

Comment (andychu) [reply core/state.py]: Oh yes, that makes sense!  I wasn't sure how similar they were

File: builtin/pure_ysh.py [hunk 2]
```diff
@@ -80,7 +41,7 @@ def Run(self, cmd_val):
             # But should there be a whitelist?
             raise error.Usage('expected a block', loc.Missing)
 
-        pairs = []  # type: List[Tuple[str, value_t]]
+        vars = {}  # type: Dict[str, value_t]
```
Comment (andychu) [inline builtin/pure_ysh.py:44]: Oh I think we might need `NewDict()` here

This is a trick/hack to preserve dict ordering in Python (underneath it uses OrderedDict)

In mycpp, our `Dict<K, V>` is defined to be ordered

(I hit this the other day too)

**Review by andychu — COMMENTED** [review]
Oh yes, good idea!

**Comment by andychu** [comment]
Awesome, thank you!  Now my question is if we should also use `ctx_Eval` for TWO places in Hay

I'll post on Zulip

**Pull request merged**